JavaScript/JavaScriptSDK.Tests/CheckinTests/TelemetryContext.tests.ts

```diff
@@ -68,6 +74,253 @@ class TelemetryContextTests extends TestClass {
+                    
+                // act
+                (<any>this._telemetryContext)._track(eventEnvelope);
+
+                // verify
+                Assert.ok(spy.calledOnce, "telemetryInitializer was called");
+                Assert.ok(eventEnvelope === spy.args[0][0]);
+
+                // teardown
+                spy.restore();
+                this._telemetryContext.onInitializeTelemetry = undefined;
+            }
+        });
+
+        this.testCase({
+            name: "TelemetryContext: onInitializeTelemetry changes the envelope props and sender gets them",
+            test: () => {
+                var nameOverride = "my unique name";
+                var eventEnvelope = this.getTestEventEnvelope();
+                Assert.notEqual(eventEnvelope.name, nameOverride);
+                var telemetryInitializer = {
+                    onInitializeTelemetry: (envelope: Microsoft.ApplicationInsights.Telemetry.Common.Envelope) => {
+                        envelope.name = nameOverride;
+                        return true;
+                    }
+                }
+
+                this._telemetryContext.onInitializeTelemetry = <any>telemetryInitializer.onInitializeTelemetry;
+                var stub = sinon.stub(this._telemetryContext._sender, "send");
+                    
+                // act
+                (<any>this._telemetryContext)._track(eventEnvelope);
+
+                // verify
+                Assert.ok(stub.calledOnce, "sender was called");
+                Assert.ok(eventEnvelope === stub.args[0][0]);
+                Assert.equal(nameOverride,
+                    (<Microsoft.ApplicationInsights.Telemetry.Common.Envelope>stub.args[0][0]).name);
+
+                // teardown
+                stub.restore();
+                this._telemetryContext.onInitializeTelemetry = undefined;
+            }
+        });
+
+        this.testCase({
+            name: "TelemetryContext: onInitializeTelemetry is null means envelope goes straight to the sender",
+            test: () => {
+                var eventEnvelope = this.getTestEventEnvelope();
+                var stub = sinon.stub(this._telemetryContext._sender, "send");
+                    
+                // act
+                (<any>this._telemetryContext)._track(eventEnvelope);
+
+                // verify
+                Assert.ok(stub.calledOnce, "sender was called");
+                Assert.ok(eventEnvelope === stub.args[0][0]);
+
+                // teardown
+                stub.restore();
+            }
+        });
+
+        this.testCase({
+            name: "TelemetryContext: onInitializeTelemetry returns false means item is not sent",
+            test: () => {
+                var eventEnvelope = this.getTestEventEnvelope();
+                var telemetryInitializer = {
+                    onInitializeTelemetry: (envelope) => {
+                        return false;
```

Comment:
As @olegsych mentioned, we are implementing chaining in .Net sdk. The reason it would be needed is when you have things like Filters where there can be more than one.